### PR TITLE
feat(build): comprehensive build system improvements with enhanced testing and code quality

### PR DESCRIPTION
**Pull request by amery**
# Summary

This PR introduces comprehensive improvements to the build system, refactors code for better maintainability and standards compliance, and significantly improves test coverage with table-driven patterns following project conventions.

## Build System Enhancements

- **Enhanced fix_whitespace.sh** with cross-platform support (SED environment variable for macOS compatibility)
- **Added comprehensive shellcheck integration** with automatic shell script linting, fallback detection, and extensive warning fixes
- **Added jq dependency validation** for dictionary generation
- **Improved coverage file organization** - merged coverage files now stored in `.tmp/coverage/` instead of root
- **Updated codecov workflow** to use new coverage file location
- **Enhanced shell script robustness** with `-r` flag for read commands, proper quoting, and security fixes

## Code Quality Improvements

- **Refactored string comparisons** to use idiomatic `s == ""` instead of `len(s) == 0`
- **Simplified type declarations** for consecutive parameters/returns with same type
- **Follow Go best practices** for string comparisons and function signatures
- **Enhanced shell script safety** with proper printf format strings and parameter expansion

## Test Coverage Improvements

- **Added comprehensive table-driven test coverage** for previously untested functions:
  - `IsErrorFn` and `IsErrorFn2` error handling functions with wrapped error support
  - `ListForEach*` family of functions for list iteration with helper functions
  - `SliceMinus` and `SliceMinusFn` slice operations with edge case coverage
  - Enhanced `Frame.SplitName` stack function coverage
- **Overall project coverage** improved from ~24% to **30.1%**
- **Created new test files**: `errors_test.go`, `lists_test.go`, plus additions to `slices_test.go` and `stack_test.go`
- **All tests follow table-driven patterns** using project conventions (S[T]() helper, subtests, comprehensive edge cases)
- **Fixed potential panic risks** in test functions with proper nil checking

## Technical Details

- Import and adapt improvements from slog's PR branch build system
- Add comprehensive shellcheck support with extensive warning fixes and disable comments
- Improve cross-platform compatibility for macOS and Linux
- Align coverage system with slog repository improvements
- All tests follow table-driven patterns with comprehensive edge case coverage
- Address CodeRabbit feedback for test consistency and error unwrapping

## Test plan

- [x] All existing tests pass
- [x] New tests provide comprehensive coverage for modified functions
- [x] Build system validates correctly with new tools
- [x] Shell scripts pass shellcheck validation with proper disable comments
- [x] Cross-platform compatibility verified
- [x] Coverage system works correctly with new file organization
- [x] Codecov integration updated and functional
- [x] No breaking changes to existing functionality
- [x] All tests follow table-driven patterns per project conventions

## Coverage Impact

- **Functions with 0% coverage now tested**: `IsErrorFn`, `IsErrorFn2`, `ListForEach*`, `SliceMinus*`
- **Overall coverage increase**: ~24% → **30.1%** (+6+ percentage points)
- **All modified functions** from this PR now have adequate test coverage
- **Enhanced test quality** with proper error unwrapping, edge cases, and table-driven patterns

## Shell Script Improvements

- **Security fixes**: Proper printf format strings, parameter expansion quoting
- **Robustness**: Read commands with `-r` flag, proper null handling
- **Documentation**: Shellcheck disable comments explaining intentional behavior
- **Compatibility**: Cross-platform sed usage, proper glob handling

<\!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Chores**
  - Added comprehensive shell script linting and improved dependency checks in the build process.
  - Updated coverage output locations and workflow configuration.
  - Refined .gitignore entries for coverage files.

- **Refactor**
  - Improved shell script compatibility and argument handling with security fixes.
  - Made minor code style improvements and used more concise function signatures.

- **Bug Fixes**
  - Enhanced shell safety and correctness in internal scripts.
  - Fixed potential panic risks in test functions.

- **Tests**
  - Introduced comprehensive table-driven unit tests for error handling, list, slice, and stack functionalities.
  - Added proper error unwrapping and edge case coverage.

<\!-- end of auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved shell script compatibility and robustness across various build scripts, including enhanced argument handling and cross-platform support.
  - Updated the Makefile to integrate shell script linting and to check for required tools before running certain build steps.
  - Adjusted coverage report paths for consistency in build outputs and CI integration.
  - Simplified and clarified code style in several internal functions and method signatures.

- **Tests**
  - Added comprehensive unit tests for error handling, list iteration, slice operations, and stack frame name splitting to ensure reliability.

- **Chores**
  - Enhanced shellcheck compliance and code clarity in build scripts.
  - Updated .gitignore to remove obsolete coverage file entries.

### CODE REVIEW

File: internal/build/gen_mk.sh [hunk 1]
```diff
@@ -82,13 +83,13 @@ GO_FILES = \$(shell find * \\
 
 EOT
 
-	while IFS=: read name dir mod deps; do
+	while IFS=: read -r name dir mod deps; do
 		files=GO_FILES_$(gen_var_name "$name")
 		filter="-e '/^\.$/d;'"
-		[ "x$dir" = "x." ] || filter="$filter -e '/^$(escape_dir "$dir")$/d;'"
+		[ "$dir" = "." ] || filter="$filter -e '/^$(escape_dir "$dir")$/d;'"
 		out_pat="$(cut -d: -f2 "$INDEX" | eval "sed $filter -e 's|$|/%|'" | tr '\n' ' ' | sed -e 's| \+$||')"
```
[the remark this answers is not in the document]
Comment (amery) [reply internal/build/gen_mk.sh:90]: safe source generated
